src/mr_allocator.rs

```diff
@@ -1,60 +1,722 @@
+    alloc: Some(RDMA_ALLOC_EXTENT_HOOK),
+    dalloc: Some(RDMA_DALLOC_EXTENT_HOOK),
+    destroy: None,
+    commit: None,
+    decommit: None,
+    purge_lazy: None,
+    purge_forced: None,
+    split: None,
+    merge: Some(RDMA_MERGE_EXTENT_HOOK),
 };
 
+lazy_static! {
+    /// Default extent hooks of jemalloc
+    static ref ORIGIN_HOOKS: extent_hooks_t = get_default_hooks_impl(DEFAULT_ARENA_INDEX);
+    /// The correspondence between extent metadata and `raw_mr`
+    #[derive(Debug)]
+    pub(crate) static ref EXTENT_TOKEN_MAP: Arc<ParkingLotMutex<BTreeMap<usize, Item>>> = Arc::new(ParkingLotMutex::new(BTreeMap::<usize, Item>::new()));
+    /// The correspondence between `arena_ind` and `ProtectionDomain`
+    pub(crate) static ref ARENA_PD_MAP: Arc<LockFreeCuckooHash<u32, Arc<ProtectionDomain>>> = Arc::new(LockFreeCuckooHash::new());
+}
+
+/// Combination between extent metadata and `raw_mr`
+#[derive(Debug)]
+pub(crate) struct Item {
+    /// addr of an extent(memory region)
+    addr: usize,
+    /// length of an extent(memory region)
+    len: usize,
+    /// Reference of RawMemoryRegion
+    raw_mr: Arc<RawMemoryRegion>,
+    /// arena index of this extent
+    arena_ind: u32,
+}
+
 /// Memory region allocator
 #[derive(Debug)]
 pub(crate) struct MrAllocator {
     /// Protection domain that holds the allocator
-    pd: Arc<ProtectionDomain>,
+    _pd: Arc<ProtectionDomain>,
+    /// Arena index
+    arena_ind: u32,
 }
 
 impl MrAllocator {
     /// Create a new MR allocator
     pub(crate) fn new(pd: Arc<ProtectionDomain>) -> Self {
-        Self { pd }
+        #[allow(clippy::expect_used)]
+        let arena_ind =
+            init_je_statics(Arc::<ProtectionDomain>::clone(&pd)).expect("init je statics failed");
+        Self { _pd: pd, arena_ind }
     }
 
     /// Allocate a MR according to the `layout`
+    #[allow(clippy::as_conversions)]
     pub(crate) fn alloc(self: &Arc<Self>, layout: &Layout) -> io::Result<LocalMr> {
-        let addr = unsafe { alloc(*layout) };
-        let access = ibv_access_flags::IBV_ACCESS_LOCAL_WRITE
-            | ibv_access_flags::IBV_ACCESS_REMOTE_WRITE
-            | ibv_access_flags::IBV_ACCESS_REMOTE_READ
-            | ibv_access_flags::IBV_ACCESS_REMOTE_ATOMIC;
-        let raw_mr = Arc::new(RawMemoryRegion::register_from_pd(
-            &self.pd,
-            addr,
-            layout.size(),
-            access,
-        )?);
-        Ok(LocalMr::new(raw_mr.addr(), raw_mr.length(), raw_mr))
+        self.alloc_from_je(layout).map_or_else(||{
+            Err(io::Error::new(io::ErrorKind::OutOfMemory, "insufficient contiguous memory was available to service the allocation request"))
+        }, |addr|{
+            let raw_mr = self.lookup_raw_mr(addr as usize);
+            Ok(LocalMr::new(addr as usize, layout.size(), raw_mr))
+        })
+    }
+
+    /// Alloc memory for RDMA operations from jemalloc
+    #[allow(clippy::as_conversions)]
```

Comment:
where is "as" used？

Reply:
aha, I replaced as with cast() but forget to remove that.